xdsl/dialects/builtin.py

```diff
@@ -1725,132 +1726,61 @@ def get_element_type(self) -> _UnrankedMemrefTypeElems:
 @irdl_attr_definition
 class DenseIntOrFPElementsAttr(TypedAttribute, ContainerType[AnyDenseElement]):
     name = "dense"
-    type: ParameterDef[
-        RankedStructure[IntegerType]
-        | RankedStructure[IndexType]
-        | RankedStructure[AnyFloat]
-    ]
-    data: ParameterDef[ArrayAttr[AnyIntegerAttr] | ArrayAttr[AnyFloatAttr]]
+    type: ParameterDef[RankedStructure[IntegerType | IndexType | AnyFloat]]
+    data: ParameterDef[BytesAttr]
+
+    @staticmethod
+    def get_format_str(
+        type: RankedStructure[IntegerType | IndexType | AnyFloat],
+    ) -> str:
+        data_len = prod(type.get_shape())
+        element_type = type.element_type
+
+        if isinstance(element_type, IndexType):
+            # store IndexType as long long in bytes object
+            format_str = "{}q"
+        elif isinstance(element_type, IntegerType):
+            if element_type.size == 1:
+                # 1 byte signed char
+                format_str = "{}b"
```

Comment:
it would be great to share code with the `ptr.py` file, if possible, potentially moving shared utilities to a third location.